legacy/include/distconv/cudnn/pooling.hpp

```diff
@@ -1,265 +1,309 @@
 #pragma once
 
-#include "distconv/util/util.hpp"
 #include "distconv/cudnn/backend.hpp"
+#include "distconv/distconv.hpp"
+#include "distconv/runtime_gpu.hpp"
 #include "distconv/tensor/halo_exchange_cuda.hpp"
-#include "distconv/tensor/halo_exchange_cuda_mpi.hpp"
 #include "distconv/tensor/halo_exchange_cuda_al.hpp"
+#include "distconv/tensor/halo_exchange_cuda_mpi.hpp"
+#include "distconv/util/util.hpp"
 #ifdef DISTCONV_HAS_P2P
-#include "distconv/tensor/halo_exchange_cuda_p2p.hpp"
 #include "distconv/tensor/halo_exchange_cuda_hybrid.hpp"
+#include "distconv/tensor/halo_exchange_cuda_p2p.hpp"
 #endif // DISTCONV_HAS_P2P
 #ifdef DISTCONV_HAS_NVSHMEM
 #include "distconv/tensor/halo_exchange_cuda_nvshmem.hpp"
 #endif // DISTCONV_HAS_NVSHMEM
 
-namespace distconv {
-namespace cudnn {
-
-static inline cudnnPoolingMode_t get_cudnn_pooling_mode(
-    const std::string &name, bool deterministic) {
-  if (name == "MAX") {
-    // This does not seem to be necessary. It's not clear what the
-    // difference of the two algorithms is.
-    if (deterministic) {
-      return CUDNN_POOLING_MAX_DETERMINISTIC;
-    } else {
-      return CUDNN_POOLING_MAX;
+namespace distconv
+{
+
+#if H2_HAS_CUDA
+namespace cudnn
+{
+static inline cudnnPoolingMode_t get_pooling_mode(const std::string& name,
+                                                  bool deterministic)
+{
+    if (name == "MAX")
+    {
+        // This does not seem to be necessary. It's not clear what the
+        // difference of the two algorithms is.
+        if (deterministic)
+        {
+            return CUDNN_POOLING_MAX_DETERMINISTIC;
+        }
+        else
+        {
+            return CUDNN_POOLING_MAX;
+        }
+    }
+    else if (name == "AVERAGE")
+    {
+        return CUDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING;
+    }
+    else if (name == "AVERAGE_NO_PAD")
+    {
+        return CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING;
+    }
+    else
+    {
+        util::PrintStreamError()
+            << "No matching pooling mode found for CUDNN: " << name << "\n";
+        std::abort();
     }
-  } else if (name == "AVERAGE") {
-    return CUDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING;
-  } else if (name == "AVERAGE_NO_PAD") {
-    return CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING;
-  } else {
-    util::PrintStreamError() << "No matching pooling mode found for CUDNN: "
-                             << name << "\n";
-    std::abort();
-  }
 }
-
 } // namespace cudnn
 
+#elif H2_HAS_ROCM
```

Comment:
@benson31 Is there a reason that this isn't split into the three files for the abstract interface and the two backends?

Reply:
Yes, most of the code is the same. It's just the function mapping strings to pooling mode that is different.